pkg/channel/distributed/controller/util/compare.go

```diff
@@ -0,0 +1,205 @@
+	oldContainerCount := len(oldDeployment.Spec.Template.Spec.Containers)
+	if oldContainerCount == 0 {
+		// This is unlikely but if it happens, replace the entire old deployment with a proper one
+		logger.Warn("Old Deployment Has No Containers - Replacing Entire Deployment")
+		return newDeployment, true
+	}
+	if len(newDeployment.Spec.Template.Spec.Containers) != 1 {
+		logger.Error("New Deployment Has Incorrect Number Of Containers And Cannot Be Used")
+		return oldDeployment, false
+	}
+
+	newContainer := &newDeployment.Spec.Template.Spec.Containers[0]
+	oldContainer := findContainer(oldDeployment, newContainer.Name)
+	if oldContainer == nil {
+		logger.Error("Old Deployment Does Not Have Same Container Name - Replacing Entire Deployment")
+		return newDeployment, true
+	}
+
+	ignoreFields := []cmp.Option{
+		// Ignore the fields in a Container struct which are not set directly by the distributed channel reconcilers
+		// and ones that are acceptable to be changed manually (such as the ImagePullPolicy)
+		cmpopts.IgnoreFields(*newContainer,
+			"Lifecycle",
+			"TerminationMessagePolicy",
+			"ImagePullPolicy",
+			"SecurityContext",
+			"StartupProbe",
+			"TerminationMessagePath",
+			"Stdin",
+			"StdinOnce",
+			"TTY"),
+		// Ignore some other fields buried inside otherwise-relevant ones, mainly "defaults that come from empty strings,"
+		// as there is no reason to restart the deployments for those changes.
+		cmpopts.IgnoreFields(corev1.ContainerPort{}, "Protocol"),         // "" -> "TCP"
+		cmpopts.IgnoreFields(corev1.ObjectFieldSelector{}, "APIVersion"), // "" -> "v1"
+		cmpopts.IgnoreFields(corev1.HTTPGetAction{}, "Scheme"),           // "" -> "HTTP" (from inside the probes; always HTTP)
+	}
+
+	containersEqual := cmp.Equal(oldContainer, newContainer, ignoreFields...)
+	if containersEqual && !metadataChanged {
+		// Nothing of interest changed, so just keep the old deployment
+		return oldDeployment, false
+	}
+
+	// Create an updated deployment from the old one, but using the new Container field
+	updatedDeployment := oldDeployment.DeepCopy()
+	if metadataChanged {
+		updatedDeployment.ObjectMeta.Labels = updatedLabels
+		updatedDeployment.Spec.Template.ObjectMeta.Annotations = updatedAnnotations
+	}
+	if !containersEqual {
+		updatedDeployment.Spec.Template.Spec.Containers[0] = *newContainer
+	}
+	return updatedDeployment, true
+}
+
+// findContainer returns the Container with the given name in a Deployment, or nil if not found
+func findContainer(deployment *appsv1.Deployment, name string) *corev1.Container {
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		if container.Name == name {
+			return &container
+		}
+	}
+	return nil
+}
+
+// CheckServiceChanged Modifies A Service With New Fields (If Necessary)
+// Returns True If Any Modifications Were Made
+func CheckServiceChanged(logger *zap.Logger, oldService, newService *corev1.Service) ([]byte, bool) {
```

Comment:
Any specific reason that this function returns a patch and then the patch is applied but in deployment the updated deployment is returned and it is updated with that?

Reply:
I am approving the PR. This was a question, not a problem with the PR.

Reply:
It was originally more similar to the CheckDeploymentChanged function, but the generation of the patch (and related error handling) was a collection of duplicated code for the two services that call CheckServiceChanged(), so I put the patch code in compare.go as well (I then separated it out into createJsonPatch() later for readability and testing purposes, so I suppose I could have made that a public function instead but it didn't occur to me at the time).

Reply:
I just realized that the question might be more "why am I using a patch and not an update," which is because for some reason Kubernetes rejects service updates that contain immutable fields (such as the "ClusterIP") _even if_ the value in the field is the same as the one in the current resource.  The patches don't contain those fields, so they cause less trouble.

Reply:
> even if the value in the field is the same as the one in the current resource. The patches don't contain those fields, so they cause less trouble.

Ok, got it.

Thanks for the detailed reply.